### ISSUE TITLE
macos: UDP server rebinds to *:port after first query

### ISSUE THREAD

**Issue by vavrusa** [miekg/dns#941]
I was looking at an odd bug report and it looks like the PacketConn server always rebinds to any address after first query. Here's a simple program:

```go
package main

import (
	"net"

	"github.com/miekg/dns"
)

func main() {
	pc, err := net.ListenPacket("udp", "127.0.0.1:1053")
	if err != nil {
		panic(err)
	}

	server := &dns.Server{PacketConn: pc}
	server.ActivateAndServe()
}
```

When I run it it binds to 127.0.0.1:1053 correctly:

```bash
$ go run bug.go &
$ lsof -p 10509 -n | grep UDP
bug     10509 vavrusa    3u    IPv4 0x52525d6e93c51cd9      0t0         UDP 127.0.0.1:remote-as
```

Cool. When I query it it answers, but the binding changes to *:1053:

```bash
$ kdig @127.0.0.1 -p 1053 +short
$ lsof -p 10509 -n | grep UDP
bug     10509 vavrusa    3u    IPv4 0x52525d6e93c51cd9      0t0         UDP *:remote-as
```

It also answers on any interface now:

```bash
$ kdig @192.168.86.110 -p 1053
;; ->>HEADER<<- opcode: QUERY; status: SERVFAIL; id: 16510
;; Flags: qr rd; QUERY: 1; ANSWER: 0; AUTHORITY: 0; ADDITIONAL: 0

;; QUESTION SECTION:
;; .                   		IN	NS

;; Received 17 B
;; Time 2019-03-17 20:03:06 PDT
;; From 192.168.86.110@1053(UDP) in 2.8 ms
```

I traced the problem to https://github.com/miekg/dns/blob/master/server.go#L326 which seems to have been added in https://github.com/miekg/dns/pull/594. The problematic part is https://github.com/miekg/dns/blob/master/udp.go#L60:

```go
// Try setting the flags for both families and ignore the errors unless they
// both error.
err6 := ipv6.NewPacketConn(conn).SetControlMessage(ipv6.FlagDst|ipv6.FlagInterface, true)
err4 := ipv4.NewPacketConn(conn).SetControlMessage(ipv4.FlagDst|ipv4.FlagInterface, true)
```

It does what it says (converts it into a raw socket). I'm not sure why was it added in the first place, but it is probably not what you want, as it opens up a DNS server using this library to any interface without user's control. Like CoreDNS:

```bash
$ cat Corefile
.:1053 {
    bind 127.0.0.1
    forward . 8.8.8.8:53
    log
}
$ coredns &
$ lsof -p $(ps ax | grep coredns | grep -v grep | awk '{print $1}') | grep UDP
coredns 16303 vavrusa    6u    IPv4 0x52525d6e9d492501      0t0         UDP localhost:remote-as
$ kdig @127.0.0.1 -p 1053 google.com +short
216.58.194.206
$ lsof -p $(ps ax | grep coredns | grep -v grep | awk '{print $1}') | grep UDP
coredns 16303 vavrusa    6u    IPv4 0x52525d6e9d492501      0t0         UDP *:remote-as
```

**Comment by tmthrgd** [comment]
This looks like the same problem as #724.

What system are you running on? (`uname -a` and `go version`)

The `SetControlMessage` lines set socket options that allow the UDP responses to be sent to the correct Ethernet adaptor.

**Comment by vavrusa** [comment]
It looks like the same issue as #724

```
$ uname -a
Darwin local 18.2.0 Darwin Kernel Version 18.2.0: Thu Dec 20 20:46:53 PST 2018; root:xnu-4903.241.1~1/RELEASE_X86_64 x86_64
$ go version
go version go1.12.1 darwin/amd64
```

**Comment by miekg** [comment]
Can we just call `SetControlMessage` without `ipv4.NewPacketConn(conn)`?

**Comment by miekg** [comment]
just to confirm doesn't happen on Linux

**Comment by tmthrgd** [comment]
The `NewPacketConn` calls don’t do anything to the underlying socket, they just return a `struct` wrapping the `net.PacketConn`.

My only guess is this is an undocumented Darwin kernel bug. It might be possible to work around it, *if* we can isolate it.

**Comment by tmthrgd** [comment]
@vavrusa Could you try running this test program I threw together for #724? [osx-oob-test.go.txt](https://github.com/miekg/dns/files/3130143/osx-oob-test.go.txt) (Rename to remove the .txt extension).

Something like this should work:
```bash
go run osx-oob-test.go &
echo test | nc -u 127.0.0.1 5053
echo test | nc -u 127.0.0.1 5053
echo test | nc -u 127.0.0.1 5053
```

You might need to adjust the flags passed to `netstat` as I don't think I ever got a chance to actually run this on an OSX machine.

**Comment by miekg** [comment]
is this is still a thing? Otherwise I'll close

**Comment by andrewtj** [comment]
@miekg it was still broken as of a couple of months back. I couldn't replicate it using native APIs.

**Comment by vavrusa** [comment]
Same problem when I run the reproducer on master

```
$ go version
go version go1.13.4 darwin/amd64
```

**Comment by andrewtj** [comment]
Does anyone want to volunteer to step through this?

**Comment by miekg** [comment]
seems noone cares enough about macos to fix this. I'll close this at some point

**Comment by denysvitali** [comment]
I think the issue is with some macOS system binaries binding on `*:53`.
In my case, I was trying to run CoreDNS on `:53`, and the first request was always successful, but the others weren't.

I digged a bit and found out the following:

macOS has a binary `/usr/sbin/mDNSResponder` that keeps on binding on TCP/53 and UDP/53 + UDP/5353.
I'm pretty sure this is used for the Bonjour service.

On top of that, in my case, Spotify is listening on 5353 too.

```
sudo lsof -nP -iTCP | grep ':53 '
mDNSRespo   205 _mdnsresponder   20u  IPv4 0x90f33fcccca090f5      0t0  TCP *:53 (LISTEN)
mDNSRespo   205 _mdnsresponder   21u  IPv6 0x90f33fcccca25c95      0t0  TCP *:53 (LISTEN)

sudo lsof -nP -iUDP | grep ':53'
mDNSRespo   205 _mdnsresponder    6u  IPv4 0x90f33fd199e59435      0t0  UDP *:5353
mDNSRespo   205 _mdnsresponder    7u  IPv6 0x90f33fd199e59745      0t0  UDP *:5353
mDNSRespo   205 _mdnsresponder   17u  IPv4 0x90f33fd199e31f75      0t0  UDP *:53
mDNSRespo   205 _mdnsresponder   19u  IPv6 0x90f33fd199e32285      0t0  UDP *:53
Spotify     696       taavide3   53u  IPv4 0x90f33fd199e337f5      0t0  UDP *:5353
Spotify     696       taavide3   88u  IPv4 0x90f33fd199e328a5      0t0  UDP *:5353
Spotify     696       taavide3  103u  IPv6 0x90f33fd199e365e5      0t0  UDP *:5353
Spotify     696       taavide3  104u  IPv6 0x90f33fd199e368f5      0t0  UDP *:5353
```

This makes only things difficult if you try to listen on `TCP/53` or `UDP/53` or `UDP/5353`.

To be able to use a custom DNS server on my mac, I did the following:

**Create an alias for the `lo0` interface:**
```
$sudo ifconfig lo0 alias 127.0.0.2
$ ping 127.0.0.2
PING 127.0.0.2 (127.0.0.2): 56 data bytes
64 bytes from 127.0.0.2: icmp_seq=0 ttl=64 time=0.259 ms
^C
--- 127.0.0.2 ping statistics ---
1 packets transmitted, 1 packets received, 0.0% packet loss
round-trip min/avg/max/stddev = 0.259/0.259/0.259/0.000 ms
```

**Configure CoreDNS to listen on `:5354`**

`Corefile`:
```
corp.example.com:5354 {
	bind 127.0.0.2
        forward . 8.8.8.8 8.8.4.4
	log
}

.:5354 {
	bind 127.0.0.2
	whoami
	forward . 1.1.1.1 1.0.0.1
	log
}
```

**Setup `pf`:**

`/etc/pf.conf`:

```
scrub-anchor "com.apple/*"
nat-anchor "com.apple/*"
rdr-anchor "com.apple/*"
rdr-anchor "com.example.dns-fix"
dummynet-anchor "com.apple/*"
anchor "com.apple/*"
load anchor "com.apple" from "/etc/pf.anchors/com.apple"
load anchor "com.example.dns-fix" from "/etc/pf.anchors/com.example.dns-fix"
```

`/etc/pf.anchors/com.example.dns-fix`:

```
rdr pass inet proto tcp from any to 127.0.0.2 port 53 -> 127.0.0.2 port 5354
rdr pass inet proto udp from any to 127.0.0.2 port 53 -> 127.0.0.2 port 5354


# Lima traffic
rdr pass inet proto tcp from 127.0.0.1 to 127.0.0.1 port 53 -> 127.0.0.2 port 5354
rdr pass inet proto udp from 127.0.0.1 to 127.0.0.1 port 53 -> 127.0.0.2 port 5354
```

**Re-apply the firewall rules**

```
$ sudo pfctl -e -f /etc/pf.conf
```


### Result

With this, I'm now able to use `127.0.0.2` as my DNS server that dynamically forwards the DNS traffic when I try to resolve `corp.example.com` to `8.8.8.8` and `8.8.4.4`. When the traffic doesn't match that domain, it gets resolved unsing `1.1.1.1` and `1.0.0.1`.

For [Lima](https://github.com/lima-vm/lima) I've added the last two rules, so that I'm able to use `192.168.5.2` (Lima's host) as a DNS server.
